### ISSUE TITLE
fields.addTaxonomy() or some other function to add taxonomy field to a list AND Taxonomy Field enum missing in FieldTypes

### ISSUE THREAD

**Issue by NewGHUser4321** [pnp/pnpjs#2590]
### Category
- [X] Enhancement
- [ ] Bug
- [ ] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: 3.10.0

Please specify what version(s) of SharePoint you are targeting: SharePoint Online

### Expected / Desired Behavior / Question
Expect a function available to add a taxonomy field to a list using PnPjs in SharePoint Online similar to `fields.addUrl(), fields.addText()` etc., Also expect Taxonomy Field `enum` available in `FieldTypes`. Do PnPjs support adding a taxonomy field to a list?

### Observed Behavior
Observed that no documentation available to add taxonomy fields to a list and also observed that `fields.addTaxonomy()` or some other function to add a taxonomy field is not available and Taxonomy Field `enum` not available in `FieldTypes` to add using `fields.add()` function

### Steps to Reproduce
Try adding a taxonomy field to a list and you wouldn't find any documentation or function.

**Comment by bcameron1231** [comment]
Hi. While there isn't a wrapper for creating the Taxonomy Field similar to the other fields types. You can create a Taxonomy field on your list by using the Schema XML. Please see [documentation](https://pnp.github.io/pnpjs/sp/fields/#create-a-field-using-an-xml-schema). This is the only way I know how to achieve this via REST.

Now, it is fairly complicated but I'll try to help below.

**Step 1**
You'll need to update a few of these values in the field schema below. Obviously the StaticName, Name, DisplayName. But also,

**ID** - Generate your own GUID
**sspId** - This is your Term Store ID
**termsetId** - The ID of the Termset the Taxonomy field should connect to
**anchorId** - Specify a term ID value here if you want a specific term to be the anchor.
```
//define schemaXML for field named 'TaxonomyFieldTest'

 const fieldSchema = `<Field Type="TaxonomyFieldType" DisplayName="TaxonomyFieldTest" ShowField="Term1033" ID="{0aa79af6-aa3c-4caa-b044-580acac36b7f}" SourceID="http://schemas.microsoft.com/sharepoint/v3" StaticName="TaxonomyFieldTest" Name="TaxonomyFieldTest" ColName="int3" RowOrdinal="0" Version="2" CustomFormatter="" EnforceUniqueValues="FALSE" Required="FALSE">
  <Default/>
  <Customization>
    <ArrayOfProperty>
      <Property>
        <Name>SspId</Name>
        <Value xmlns:q1="http://www.w3.org/2001/XMLSchema" xmlns:p4="http://www.w3.org/2001/XMLSchema-instance" p4:type="q1:string">1141fa29-9215-4811-a77e-2d8311ccc808</Value>
      </Property>
      <Property>
        <Name>GroupId</Name>
      </Property>
      <Property>
        <Name>TermSetId</Name>
        <Value xmlns:q2="http://www.w3.org/2001/XMLSchema" xmlns:p4="http://www.w3.org/2001/XMLSchema-instance" p4:type="q2:string">df373652-2b1a-4b77-befd-1054d183ef36</Value>
      </Property>
      <Property>
        <Name>AnchorId</Name>
        <Value xmlns:q3="http://www.w3.org/2001/XMLSchema" xmlns:p4="http://www.w3.org/2001/XMLSchema-instance" p4:type="q3:string">00000000-0000-0000-0000-000000000000</Value>
      </Property>
      <Property>
        <Name>UserCreated</Name>
        <Value xmlns:q4="http://www.w3.org/2001/XMLSchema" xmlns:p4="http://www.w3.org/2001/XMLSchema-instance" p4:type="q4:boolean">false</Value>
      </Property>
      <Property>
        <Name>Open</Name>
        <Value xmlns:q5="http://www.w3.org/2001/XMLSchema" xmlns:p4="http://www.w3.org/2001/XMLSchema-instance" p4:type="q5:boolean">false</Value>
      </Property>
      <Property>
        <Name>TextField</Name>
        <Value xmlns:q6="http://www.w3.org/2001/XMLSchema" xmlns:p4="http://www.w3.org/2001/XMLSchema-instance" p4:type="q6:string">{4bb86746-f0a1-412e-bc7d-aba1f0eb14ad}</Value>
      </Property>
      <Property>
        <Name>IsPathRendered</Name>
        <Value xmlns:q7="http://www.w3.org/2001/XMLSchema" xmlns:p4="http://www.w3.org/2001/XMLSchema-instance" p4:type="q7:boolean">false</Value>
      </Property>
      <Property>
        <Name>IsKeyword</Name>
        <Value xmlns:q8="http://www.w3.org/2001/XMLSchema" xmlns:p4="http://www.w3.org/2001/XMLSchema-instance" p4:type="q8:boolean">false</Value>
      </Property>
      <Property>
        <Name>TargetTemplate</Name>
      </Property>
      <Property>
        <Name>CreateValuesInEditForm</Name>
        <Value xmlns:q9="http://www.w3.org/2001/XMLSchema" xmlns:p4="http://www.w3.org/2001/XMLSchema-instance" p4:type="q9:boolean">false</Value>
      </Property>
      <Property>
        <Name>IsDocTagsEnabled</Name>
      </Property>
      <Property>
        <Name>FilterAssemblyStrongName</Name>
        <Value xmlns:q10="http://www.w3.org/2001/XMLSchema" xmlns:p4="http://www.w3.org/2001/XMLSchema-instance" p4:type="q10:string">Microsoft.SharePoint.Taxonomy, Version=16.0.0.0, Culture=neutral, PublicKeyToken=71e9bce111e9429c</Value>
      </Property>
      <Property>
        <Name>FilterClassName</Name>
        <Value xmlns:q11="http://www.w3.org/2001/XMLSchema" xmlns:p4="http://www.w3.org/2001/XMLSchema-instance" p4:type="q11:string">Microsoft.SharePoint.Taxonomy.TaxonomyField</Value>
      </Property>
      <Property>
        <Name>FilterMethodName</Name>
        <Value xmlns:q12="http://www.w3.org/2001/XMLSchema" xmlns:p4="http://www.w3.org/2001/XMLSchema-instance" p4:type="q12:string">GetFilteringHtml</Value>
      </Property>
      <Property>
        <Name>FilterJavascriptProperty</Name>
        <Value xmlns:q13="http://www.w3.org/2001/XMLSchema" xmlns:p4="http://www.w3.org/2001/XMLSchema-instance" p4:type="q13:string">FilteringJavascript</Value>
      </Property>
    </ArrayOfProperty>
  </Customization>
</Field>`

// create the new field in the list 'Test'
const field : IFieldAddResult= await sp.web.lists.getByTitle("Test").fields.createFieldAsXml(fieldSchema);
```

**Step 2**
This step becomes a bit more convoluted. If you want the ability to use search (I'd recommend this either way), you will need to deploy an additional field onto your list. SharePoint handles this for you when you create the field type from the GUI.  This field is a "Note" field. Without this field, SharePoint will not be able to create Managed or Crawled Properties for your Taxonomy Column, thus it will not be indexed.

It is a hidden Note field, with a internal name equal to your Taxonomy Field but has the value "TaxHTField" or "_0" appended to it. I would recommend the TaxHTField.

**ID** - Generate your own GUID
Update the DisplayName, StaticName, Name to be the Internal Name of your Taxonomy Field + TaxHTField
```
const noteFieldSchema = `<Field ID="{44a5d269-ce40-4a8f-b0d6-6b7bcb95ff71}" SourceID="http://schemas.microsoft.com/sharepoint/v3" Type="Note" Name="TaxonomyFieldTest" StaticName="TaxonomyFieldTestTaxHTField" DisplayName="TaxonomyFieldTestTaxHTField" ShowInViewForms="FALSE" Required="FALSE" Hidden="TRUE" CanToggleHidden="TRUE" />`

const noteField = await sp.web.lists.getByTitle("Test").fields.createFieldAsXml(noteFieldSchema);
```

I hope this at least unblocks you going forward.

**Comment by NewGHUser4321** [comment]
Thank you @bcameron1231 for the detailed instructions. I did exactly the same using the Schema XML just after I created this post and it seems to work fine. I did not do the **Step 2** though. SharePoint created the hidden field automatically for me and I also used that field when adding list items to the list.

**Comment by bcameron1231** [comment]
Perfect!  Glad you got it working! Thank you.

**Comment by patrick-rodgers** [comment]
Closing this issue as answered. If you have additional questions or we did not answer your question, please open a new issue, ref this issue, and provide any additional details available. Thank you!
